src/projection.jl

```diff
@@ -52,93 +128,197 @@ function ProjectTo(x::T) where {T<:Union{<:Tuple,NamedTuple}}
-function (project::ProjectTo{<:Transpose})(dx::AbstractMatrix)
-    return transpose(project.parent(transpose(dx)))
-end
-(project::ProjectTo{<:Transpose})(dx::AbstractZero) = transpose(project.parent(dx))
-
-# Adjoint
-ProjectTo(x::T) where {T<:Adjoint} = ProjectTo{T}(; parent=ProjectTo(parent(x)))
-(project::ProjectTo{<:Adjoint})(dx::AbstractMatrix) = adjoint(project.parent(adjoint(dx)))
-(project::ProjectTo{<:Adjoint})(dx::AbstractZero) = adjoint(project.parent(dx))
-
-# PermutedDimsArray
-ProjectTo(x::P) where {P<:PermutedDimsArray} = ProjectTo{P}(; parent=ProjectTo(parent(x)))
-function (project::ProjectTo{<:PermutedDimsArray{T,N,perm,iperm,AA}})(
-    dx::AbstractArray
-) where {T,N,perm,iperm,AA}
-    return PermutedDimsArray{T,N,perm,iperm,AA}(permutedims(project.parent(dx), perm))
+# Weird
+ProjectTo(x::Bidiagonal) = generic_projectto(x) # not sure!
+function (project::ProjectTo{<:Bidiagonal})(dx::AbstractMatrix)
+    uplo = LinearAlgebra.sym_uplo(project.uplo.value)
+    dv = project.dv(diag(dx))
+    ev = project.ev(uplo === :U ? diag(dx, 1) : diag(dx, -1))
+    Bidiagonal(dv, ev, uplo)
 end
-function (project::ProjectTo{P})(dx::AbstractZero) where {P<:PermutedDimsArray}
-    return P(project.parent(dx))
+
+#=
+
+x = LinearAlgebra.Tridiagonal(rand(4,4))
+backing(x) # UndefRefError: access to undefined reference
+
+=#
+
+
+# # SubArray
+# ProjectTo(x::T) where {T<:SubArray} = ProjectTo(copy(x))  # don't project on to a view, but onto matching copy
+
+# Sparse
+
+# using SparseArrays
+
+
+
+export proj_rrule
+"""
+    proj_rrule(f, args...)
+
+This calls the corresponding `rrule`, but automatically
+applies ProjectTo to all arguments.
+
+```jldoctest
+julia> using ChainRules
+
+julia> x, y = Diagonal(rand(3)), rand(3,3);
+
+julia> z, bk = rrule(*, x, y);
+
+julia> bk(z .* 100)[2] |> unthunk
+3×3 Matrix{Float64}:
+  8.04749   20.4969   16.3239
+ 39.2846   184.577   128.653
+ 11.1317    45.7744   32.7681
+
+julia> z2, bk2 = proj_rrule(*, x, y);
+
+julia> bk2(z .* 100)[2] |> unthunk  # now preserves the subspace
+3×3 Diagonal{Float64, Vector{Float64}}:
+ 8.04749     ⋅       ⋅ 
+  ⋅       184.577    ⋅ 
+  ⋅          ⋅     32.7681
+
+julia> @btime rrule(*, \$x, \$y)[2](\$z)[2]|>unthunk;
+  66.777 ns (2 allocations: 256 bytes)
+
+julia> @btime proj_rrule(*, \$x, \$y)[2](\$z)[2]|>unthunk;  # still not ideal, makes & throws away Matrix
+  117.798 ns (4 allocations: 416 bytes)
+
+julia> _, bk3 = rrule(*, 1, 2+3im)
+(2 + 3im, ChainRules.var"#times_pullback#1114"{Int64, Complex{Int64}}(1, 2 + 3im))
+
+julia> bk3(4+5im)
+(NoTangent(), 23 - 2im, 4 + 5im)
+
+julia> _, bk4 = proj_rrule(*, 1, 2+3im);
+
+julia> bk4(4+5im)  # works for scalars too
+(NoTangent(), 23.0, 4.0 + 5.0im)
+
+```
+"""
+function proj_rrule(f, args...)
+    ps = map(ProjectTo, args)
+    y, back = rrule(f, args...)
+    function proj_back(dy)
+        res = back(dy)
+        (first(res), map((p,dx) -> p(dx), ps, Base.tail(res))...)
```

Comment:
Why not project the function object also?

(Aside: TIL
```
julia> @ballocated tuple(1, map(identity, $(1,2,3,4))...)
0
```
)

Reply:
Because my head hurt when I tried to think about that. Very much a first sketch. 

I agree in principle it should apply to all. Whether it can usefully go inside functions which close over parameters... would have to invent some examples and have a look.

Reply:
Oh the problem is we don't have tangent types.
because ProjectTo's constructor is sensibly defined assuming the input is a for tangent type.
It gets really hard to define it if you need to workout tangent types from primal types yourself.

Which in general is an argument against applying it to *everything* in this way.
And why we were doing it inside rules where we can apply it when it make sense.

Reply:
I think this problem makes actually defining this `proj_rrule` a non-starter.
We would need to know a canonical tangent type for *everything*, and be able to generate that, just so we could feed to to `ProjectTo`.
Even things for which the answer in the rrule we are calling at least, is always going to be `NoTangent()` or `ZeroTangent()`

Reply:
I don't know, I'm not sure I follow. I thought this was literally the idea, that ProjectTo should be able digest anything. On fields whose tangents will always be Zero, only the trivial method of `(::ProjectTo)` will be called, so you need never find out whether it can reconstruct them. 

See also my comment below, there are two ways this could be structured. One of them is much more opt-in, as a way to add projections globally to a rule you are writing.

Reply:
we *could* do this by allowing `Project(x::P)` for a primal type `x` be  `Project{Tangent{P}}` if we don't know better.
They will either give us that as an output, or they will give us a struct that is a valid tangent type which we can decompose into that, or they will give us a `AbstractZero` and we don't have to care.